[BUG] throw ClassCastException for Pulsar ProtobufNative Schema with Flink  SQL

**Describe the bug**

Using ProtobufNative to decode the protobuf map type is throwing an `ClassCastException` exception.

**To Reproduce**

```
Flink SQL> select * from `test-pb-native-1`;
[ERROR] Could not execute SQL statement. Reason:
java.lang.ClassCastException: java.util.Collections$UnmodifiableRandomAccessList cannot be cast to java.util.Map
	at org.apache.flink.formats.protobufnative.PulsarProtobufToRowDataConverters.lambda$createMapConverter$bdab2584$1(PulsarProtobufToRowDataConverters.java:144)
	at org.apache.flink.formats.protobufnative.PulsarProtobufToRowDataConverters.lambda$createNullableConverter$bfe4cab$1(PulsarProtobufToRowDataConverters.java:77)
	at org.apache.flink.formats.protobufnative.PulsarProtobufToRowDataConverters.lambda$createRowConverter$52b2b307$1(PulsarProtobufToRowDataConverters.java:64)
	at org.apache.flink.formats.protobufnative.PulsarProtobufNativeRowDataDeserializationSchema.deserialize(PulsarProtobufNativeRowDataDeserializationSchema.java:63)
	at org.apache.flink.formats.protobufnative.PulsarProtobufNativeRowDataDeserializationSchema.deserialize(PulsarProtobufNativeRowDataDeserializationSchema.java:30)
	at org.apache.flink.api.common.serialization.DeserializationSchema.deserialize(DeserializationSchema.java:82)
	at org.apache.flink.streaming.util.serialization.ThreadSafeDeserializationSchema.deserialize(ThreadSafeDeserializationSchema.java:53)
	at org.apache.flink.streaming.connectors.pulsar.table.DynamicPulsarDeserializationSchema.deserialize(DynamicPulsarDeserializationSchema.java:120)
	at org.apache.flink.streaming.connectors.pulsar.table.DynamicPulsarDeserializationSchema.deserialize(DynamicPulsarDeserializationSchema.java:111)
	at org.apache.flink.streaming.connectors.pulsar.table.DynamicPulsarDeserializationSchema.deserialize(DynamicPulsarDeserializationSchema.java:42)
	at org.apache.flink.streaming.connectors.pulsar.internal.ReaderThread.emitRecord(ReaderThread.java:171)
	at org.apache.flink.streaming.connectors.pulsar.internal.ReaderThread.run(ReaderThread.java:110)

```

**Expected behavior**
A clear and concise description of what you expected to happen.

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Additional context**
Add any other context about the problem here.

--- a/pulsar-flink-connector/src/main/java/org/apache/flink/formats/protobufnative/PulsarProtobufToRowDataConverters.java
+++ b/pulsar-flink-connector/src/main/java/org/apache/flink/formats/protobufnative/PulsarProtobufToRowDataConverters.java
@@ -25,11 +25,13 @@ import org.apache.flink.table.types.logical.RowType;
 import org.apache.flink.table.types.logical.utils.LogicalTypeUtils;
 
 import com.google.protobuf.ByteString;
+import com.google.protobuf.Descriptors;
 import com.google.protobuf.DynamicMessage;
 import com.google.protobuf.EnumValue;
 
 import java.io.Serializable;
 import java.lang.reflect.Array;
+import java.util.Collection;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
@@ -136,17 +138,36 @@ public class PulsarProtobufToRowDataConverters {
     private static PulsarProtobufToRowDataConverters.ProtobufToRowDataConverter createMapConverter(MapType type) {
 
         final PulsarProtobufToRowDataConverters.ProtobufToRowDataConverter keyConverter =
-                createConverter(type.getKeyType());
+                createNullableConverter(type.getKeyType());
         final PulsarProtobufToRowDataConverters.ProtobufToRowDataConverter valueConverter =
                 createNullableConverter(type.getValueType());
 
         return object -> {
-            final Map<?, ?> map = (Map<?, ?>) object;
             Map<Object, Object> result = new HashMap<>();
-            for (Map.Entry<?, ?> entry : map.entrySet()) {
-                Object key = keyConverter.convert(entry.getKey());
-                Object value = valueConverter.convert(entry.getValue());
-                result.put(key, value);
+            if (object instanceof Collection) {
+                final Collection<DynamicMessage> messages = (Collection<DynamicMessage>) object;
+                for (DynamicMessage message : messages) {
+                    final Map<Descriptors.FieldDescriptor, Object> allFields = message.getAllFields();
+                    Object key = null;
+                    Object value = null;
+                    for (Map.Entry<Descriptors.FieldDescriptor, Object> objectEntry : allFields.entrySet()) {
+                        final Descriptors.FieldDescriptor fieldDescriptor = objectEntry.getKey();
+                        if ("key".equals(fieldDescriptor.getJsonName())) {
+                            key = keyConverter.convert(objectEntry.getValue());
+                        }
+                        if ("value".equals(fieldDescriptor.getJsonName())) {
+                            value = valueConverter.convert(objectEntry.getValue());
+                        }
+                    }
+                    result.put(key, value);
+                }
+            } else {
+                final Map<?, ?> map = (Map<?, ?>) object;
+                for (Map.Entry<?, ?> entry : map.entrySet()) {
+                    Object key = keyConverter.convert(entry.getKey());
+                    Object value = valueConverter.convert(entry.getValue());
+                    result.put(key, value);
+                }
             }
             return new GenericMapData(result);
         };

--- a/pulsar-flink-connector/src/main/java/org/apache/flink/formats/protobufnative/PulsarProtobufToRowDataConverters.java
+++ b/pulsar-flink-connector/src/main/java/org/apache/flink/formats/protobufnative/PulsarProtobufToRowDataConverters.java
@@ -25,7 +25,6 @@ import org.apache.flink.table.types.logical.RowType;
 import org.apache.flink.table.types.logical.utils.LogicalTypeUtils;
 
 import com.google.protobuf.ByteString;
-import com.google.protobuf.Descriptors;
 import com.google.protobuf.DynamicMessage;
 import com.google.protobuf.EnumValue;
 
@@ -37,6 +36,9 @@ import java.util.List;
 import java.util.Map;
 
 public class PulsarProtobufToRowDataConverters {
+
+    protected static final String PROTOBUF_MAP_KEY_NAME = "key";
+    protected static final String PROTOBUF_MAP_VALUE_NAME = "value";
 
     /**
      * Runtime converter that converts protobuf data structures into objects of Flink Table & SQL
@@ -144,30 +146,15 @@ public class PulsarProtobufToRowDataConverters {
 
         return object -> {
             Map<Object, Object> result = new HashMap<>();
-            if (object instanceof Collection) {
-                final Collection<DynamicMessage> messages = (Collection<DynamicMessage>) object;
-                for (DynamicMessage message : messages) {
-                    final Map<Descriptors.FieldDescriptor, Object> allFields = message.getAllFields();
-                    Object key = null;
-                    Object value = null;
-                    for (Map.Entry<Descriptors.FieldDescriptor, Object> objectEntry : allFields.entrySet()) {
-                        final Descriptors.FieldDescriptor fieldDescriptor = objectEntry.getKey();
-                        if ("key".equals(fieldDescriptor.getJsonName())) {
-                            key = keyConverter.convert(objectEntry.getValue());
-                        }
-                        if ("value".equals(fieldDescriptor.getJsonName())) {
-                            value = valueConverter.convert(objectEntry.getValue());
-                        }
-                    }
-                    result.put(key, value);
-                }
-            } else {
-                final Map<?, ?> map = (Map<?, ?>) object;
-                for (Map.Entry<?, ?> entry : map.entrySet()) {
-                    Object key = keyConverter.convert(entry.getKey());
-                    Object value = valueConverter.convert(entry.getValue());
-                    result.put(key, value);
-                }
+            final Collection<DynamicMessage> messages = (Collection<DynamicMessage>) object;
+            for (DynamicMessage message : messages) {
+                Object key = keyConverter.convert(
+                        message.getField(message.getDescriptorForType().findFieldByName(PROTOBUF_MAP_KEY_NAME))
+                );
+                Object value = valueConverter.convert(
+                        message.getField(message.getDescriptorForType().findFieldByName(PROTOBUF_MAP_VALUE_NAME))
+                );
+                result.put(key, value);
             }
             return new GenericMapData(result);
         };

--- a/pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/internal/SimpleSchemaTranslator.java
+++ b/pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/internal/SimpleSchemaTranslator.java
@@ -336,7 +336,7 @@ public class SimpleSchemaTranslator extends SchemaTranslator {
         }
     }
 
-    private static DataType proto2SqlType(Descriptors.Descriptor descriptor) throws
+    public static DataType proto2SqlType(Descriptors.Descriptor descriptor) throws
             IncompatibleSchemaException {
         List<DataTypes.Field> fields = new ArrayList<>();
         List<Descriptors.FieldDescriptor> protoFields = descriptor.getFields();

--- a/pulsar-flink-connector/src/test/java/org/apache/flink/formats/protobufnative/PulsarProtobufNativeRowDataDeserializationSchemaTest.java
+++ b/pulsar-flink-connector/src/test/java/org/apache/flink/formats/protobufnative/PulsarProtobufNativeRowDataDeserializationSchemaTest.java
@@ -14,35 +14,61 @@
 
 package org.apache.flink.formats.protobufnative;
 
-import org.apache.flink.formats.protobuf.PbRowTypeInformation;
-import org.apache.flink.streaming.connectors.pulsar.SchemaData;
 import org.apache.flink.table.data.RowData;
 import org.apache.flink.table.types.logical.RowType;
 
+import com.google.protobuf.Descriptors;
+import org.apache.pulsar.client.impl.schema.ProtobufNativeSchemaUtils;
+import org.junit.Before;
 import org.junit.Test;
 
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+
+import static org.apache.flink.streaming.connectors.pulsar.internal.SimpleSchemaTranslator.proto2SqlType;
 import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertFalse;
 
 public class PulsarProtobufNativeRowDataDeserializationSchemaTest {
 
+    public byte[] protobufData;
+
+    public Descriptors.Descriptor descriptor;
+
+    @Before
+    public void setUp() throws Exception {
+        String schema =
+                "{\"fileDescriptorSet\":\"CtcGCg5wYjNfdGVzdC5wcm90bxInb3JnLmFwYWNoZS5mbGluay5mb3JtYXRzLnByb3RvYnVmLnByb3RvIuIFCgdQYjNUZXN0EgkKAWEYASABKAUSCQoBYhgCIAEoAxIJCgFjGAMgASgJEgkKAWQYBCABKAISCQoBZRgFIAEoARJCCgFmGAYgASgOMjcub3JnLmFwYWNoZS5mbGluay5mb3JtYXRzLnByb3RvYnVmLnByb3RvLlBiM1Rlc3QuQ29ycHVzEkwKAWcYByABKAsyQS5vcmcuYXBhY2hlLmZsaW5rLmZvcm1hdHMucHJvdG9idWYucHJvdG8uUGIzVGVzdC5Jbm5lck1lc3NhZ2VUZXN0EkwKAWgYCCADKAsyQS5vcmcuYXBhY2hlLmZsaW5rLmZvcm1hdHMucHJvdG9idWYucHJvdG8uUGIzVGVzdC5Jbm5lck1lc3NhZ2VUZXN0EgkKAWkYCSABKAwSSAoEbWFwMRgKIAMoCzI6Lm9yZy5hcGFjaGUuZmxpbmsuZm9ybWF0cy5wcm90b2J1Zi5wcm90by5QYjNUZXN0Lk1hcDFFbnRyeRJICgRtYXAyGAsgAygLMjoub3JnLmFwYWNoZS5mbGluay5mb3JtYXRzLnByb3RvYnVmLnByb3RvLlBiM1Rlc3QuTWFwMkVudHJ5GisKCU1hcDFFbnRyeRILCgNrZXkYASABKAkSDQoFdmFsdWUYAiABKAk6AjgBGm4KCU1hcDJFbnRyeRILCgNrZXkYASABKAkSUAoFdmFsdWUYAiABKAsyQS5vcmcuYXBhY2hlLmZsaW5rLmZvcm1hdHMucHJvdG9idWYucHJvdG8uUGIzVGVzdC5Jbm5lck1lc3NhZ2VUZXN0OgI4ARooChBJbm5lck1lc3NhZ2VUZXN0EgkKAWEYASABKAUSCQoBYhgCIAEoAyJaCgZDb3JwdXMSDQoJVU5JVkVSU0FMEAASBwoDV0VCEAESCgoGSU1BR0VTEAISCQoFTE9DQUwQAxIICgRORVdTEAQSDAoIUFJPRFVDVFMQBRIJCgVWSURFTxAHQi8KK29yZy5hcGFjaGUuZmxpbmsuZm9ybWF0cy5wcm90b2J1Zi50ZXN0cHJvdG9QAWIGcHJvdG8z\",\"rootMessageTypeName\":\"org.apache.flink.formats.protobuf.proto.Pb3Test\",\"rootFileDescriptorName\":\"pb3_test.proto\"}";
+        descriptor = ProtobufNativeSchemaUtils.deserialize(schema.getBytes(StandardCharsets.UTF_8));
+        protobufData = Base64.getDecoder().decode(
+                "CGMQARoEdGVzdCXNzIw/KcP1KFyPwvE/MAE6BAhjEGNCBAhjEGNCBAhjEGNKATFSCgoDMTExEgMyMjJSDAoEMTExMhIEMjIyM1oLCgMxMTESBAhjEGNaDAoEMTExMhIECGMQYw==");
+    }
+
     @Test
     public void deserialize() throws Exception {
-        RowType rowType = PbRowTypeInformation.generateRowType(SchemaData.descriptor);
-        PulsarProtobufNativeRowDataDeserializationSchema deserializationSchema = new PulsarProtobufNativeRowDataDeserializationSchema(() -> SchemaData.descriptor, rowType);
+        RowType rowType = (RowType) proto2SqlType(descriptor).getLogicalType();
+        PulsarProtobufNativeRowDataDeserializationSchema deserializationSchema =
+                new PulsarProtobufNativeRowDataDeserializationSchema(() -> descriptor, rowType);
         deserializationSchema.open(null);
 
-        final RowData rowData = deserializationSchema.deserialize(SchemaData.protobufData);
+        final RowData rowData = deserializationSchema.deserialize(protobufData);
         RowData newRowData = rowData;
-        assertEquals(9, newRowData.getArity());
-        assertEquals(1, newRowData.getInt(0));
-        assertEquals(2L, newRowData.getLong(1));
-        assertFalse((boolean) newRowData.getBoolean(2));
-        assertEquals(Float.valueOf(0.1f), Float.valueOf(newRowData.getFloat(3)));
-        assertEquals(Double.valueOf(0.01d), Double.valueOf(newRowData.getDouble(4)));
-        assertEquals("haha", newRowData.getString(5).toString());
-        assertEquals(1, (newRowData.getBinary(6))[0]);
-        assertEquals("IMAGES", newRowData.getString(7).toString());
-        assertEquals(1, newRowData.getInt(8));
+        assertEquals(11, newRowData.getArity());
+        assertEquals(99, newRowData.getInt(0));
+        assertEquals(1L, newRowData.getLong(1));
+        assertEquals("test", newRowData.getString(2).toString());
+        assertEquals(Float.valueOf(1.1f), Float.valueOf(newRowData.getFloat(3)));
+        assertEquals(Double.valueOf(1.11d), Double.valueOf(newRowData.getDouble(4)));
+        assertEquals("WEB", newRowData.getString(5).toString());
+        assertEquals("+I(99,99)", newRowData.getRow(6, 2).toString());
+        assertEquals(2, newRowData.getArray(7).size());
+        assertEquals(99, newRowData.getArray(7).getRow(0, 1).getInt(0));
+        assertEquals(99, newRowData.getArray(7).getRow(1, 1).getInt(0));
+        assertEquals(49, newRowData.getBinary(8)[0]);
+        assertEquals(2, newRowData.getMap(9).size());
+        assertEquals("1112", newRowData.getMap(9).keyArray().getString(0).toString());
+        assertEquals("2223", newRowData.getMap(9).valueArray().getString(0).toString());
+        assertEquals(2, newRowData.getMap(10).size());
+        assertEquals("1112", newRowData.getMap(10).keyArray().getString(0).toString());
+        assertEquals("+I(99,99)", newRowData.getMap(10).valueArray().getRow(0, 2).toString());
     }
 }

